include/sacn/cpp/merge_receiver.h

```diff
@@ -0,0 +1,370 @@
+ * \return #kEtcPalErrOk: Receiver created successful.
+ * \return #kEtcPalErrInvalid: Invalid parameter provided.
+ * \return #kEtcPalErrNotInit: Module not initialized.
+ * \return #kEtcPalErrExists: A receiver already exists which is listening on the specified universe.
+ * \return #kEtcPalErrNoMem: No room to allocate memory for this receiver.
+ * \return #kEtcPalErrNoNetints: No network interfaces were found on the system.
+ * \return #kEtcPalErrNotFound: A network interface ID given was not found on the system.
+ * \return #kEtcPalErrSys: An internal library or system call error occurred.
+ */
+inline etcpal::Error MergeReceiver::Startup(const Settings& settings, NotifyHandler& notify_handler)
+{
+  SacnMergeReceiverConfig config = TranslateConfig(settings, notify_handler);
+  return sacn_merge_receiver_create(&config, &handle_);
+}
+
+/*!
+ * \brief Stop listening for sACN data on a universe.
+ *
+ * Tears down the receiver and any sources currently being tracked on the receiver's universe.
+ * Stops listening for sACN on that universe.
+ */
+inline void MergeReceiver::Shutdown()
+{
+  // We'll be ignoring shutdown errors for now
+  sacn_merge_receiver_destroy(handle_);
+  handle_ = kInvalidHandle;
+}
+
+/*!
+ * \brief Get the universe this class is listening to.
+ *
+ * \return If valid, the value is the universe id.  Otherwise, this is the underlying error the C library call returned.
+ */
+etcpal::Expected<uint16_t> MergeReceiver::GetUniverse() const
+{
+  uint16_t result = 0;
+  etcpal_error_t err = sacn_merge_receiver_get_universe(handle_, &result);
+  if (err == kEtcPalErrOk)
+    return result;
+  else
+    return err;
+}
+
+/*!
+ * \brief Change the universe this class is listening to.
+ *
+ * An sACN receiver can only listen on one universe at a time. After this call completes successfully, the receiver is
+ * in a sampling period for the new universe and will provide HandleSourcesFound() callse when appropriate.
+ * If this call fails, the caller must call Shutdown() on this class, because it may be in an invalid state.
+ *
+ * \param[in] new_universe_id New universe number that this receiver should listen to.
+ * \return #kEtcPalErrOk: Universe changed successfully.
+ * \return #kEtcPalErrInvalid: Invalid parameter provided.
+ * \return #kEtcPalErrNotInit: Module not initialized.
+ * \return #kEtcPalErrExists: A receiver already exists which is listening on the specified new universe.
+ * \return #kEtcPalErrNotFound: Handle does not correspond to a valid receiver.
+ * \return #kEtcPalErrSys: An internal library or system call error occurred.
+ */
+inline etcpal::Error MergeReceiver::ChangeUniverse(uint16_t new_universe_id)
+{
+  return sacn_merge_receiver_change_universe(handle_, new_universe_id);
+}
+
+/*!
+ * \brief Resets the underlying network sockets and packet receipt state for this class..
+ *
+ * This is typically used when the application detects that the list of networking interfaces has changed.
```

Comment:
I'm not convinced we should handle it internally. It's one thing for the RDMnet broker to handle this as part of it's binary, but this library will be built into a larger application, and the application may want to control when this happens in the larger ecosystem.

It also opens a can of worms when the receiver was created for a specific set of NICs and that set  needs to change over time due to wireless NICs getting disabled/enabled.  This started the potential AsyncSocket down a nasty path of needing to track NIC state, monitor for NICs getting added & deleted, and callbacks that notify library and application code.  We never completed it.

Reply:
The library will be built into a larger application, but the library would only be responsible for handling network change with regard to its own functionality, not the functionality of the larger application. I do see your point about controlling when this happens.

The reason I brought this up is that handling network changes gracefully is one of the top feature requests we get for our libraries from engineers and TPMs.

Reply:
This a top request I was getting as well.  My thought was that this function helped resolve that, as the app can tell us when to do it over what interfaces, and the library would gracefully change.  The list of NICs changing over time particularly worried me.
Do we have network change detection built into EtcPal already?

Reply:
It's proposed but not implemented yet. It will be polling-based (etcpal_netint_refresh_interfaces(), will return true if the interfaces have changed, false if they are the same).

Reply:
How will it handle NICs that disappear?  Reserve the netintid for later re-use if that NIC appears?  Or am I missing the point -- what is used as the NIC identifier here?

Reply:
I would think that if the library has been explicitly told to listen on an interface, and that interface goes away, the library should deallocate and not reopen its socket on that NIC.

I see from exploring the problem in my head, based on your comment, that there are a lot of corner cases. I'm more open to this than I was initially.

Reply:
Well, think about it and we can talk more.  This function was my way of taking a stab at it without adding too much complexity/overhead in the library.  The "listen on all interfaces" case is easy and most likely to be the embedded use case, but we need to support more than that on the bigger OSs.